.github/workflows/R-CMD-check.yaml

```diff
@@ -27,92 +23,74 @@ jobs:
+          - {os: ubuntu-20.04,   r: 'devel', http-user-agent: 'release'}
+          - {os: ubuntu-20.04,   r: 'release'}
+          - {os: ubuntu-20.04,   r: 'release', custom: 'no-cpp11test'}
+          - {os: ubuntu-20.04,   r: 'release', custom: 'gcc 4.8'}
+          - {os: ubuntu-20.04,   r: 'oldrel-1'}
+          - {os: ubuntu-20.04,   r: 'oldrel-2'}
+          - {os: ubuntu-20.04,   r: 'oldrel-3'}
+          - {os: ubuntu-20.04,   r: 'oldrel-4'}
 
     env:
-      RSPM: ${{ matrix.config.rspm }}
       GITHUB_PAT: ${{ secrets.GITHUB_TOKEN }}
+      R_KEEP_PKG_SOURCE: yes
 
     steps:
       - uses: actions/checkout@v2
 
-      - uses: r-lib/actions/setup-r@v1
-        id: install-r
+      - uses: r-lib/actions/setup-pandoc@v2
+
+      - uses: r-lib/actions/setup-r@v2
         with:
           r-version: ${{ matrix.config.r }}
           http-user-agent: ${{ matrix.config.http-user-agent }}
-
-      - uses: r-lib/actions/setup-pandoc@v1
-
-      - name: Install pak and query dependencies
-        run: |
-          install.packages("pak", repos = "https://r-lib.github.io/p/pak/dev/")
-          saveRDS(pak::pkg_deps("local::.", dependencies = TRUE), ".github/r-depends.rds")
-        shell: Rscript {0}
-
-      - name: Restore R package cache
-        uses: actions/cache@v2
-        with:
-          path: |
-            ${{ env.R_LIBS_USER }}/*
-            !${{ env.R_LIBS_USER }}/pak
-          key: ${{ matrix.config.os }}-${{ steps.install-r.outputs.installed-r-version }}-1-${{ hashFiles('.github/r-depends.rds') }}
-          restore-keys: ${{ matrix.config.os }}-${{ steps.install-r.outputs.installed-r-version }}-1-
-
-      - name: Install system dependencies
-        if: runner.os == 'Linux'
-        run: |
-          pak::local_system_requirements(execute = TRUE)
-          pak::pkg_system_requirements("rcmdcheck", execute = TRUE)
-        shell: Rscript {0}
+          use-public-rspm: true
 
       - name : Install compiler version
+        # We check on this old compiler specifically to support CentOS 7,
+        # which uses this gcc version. RStudio products support CentOS 7 through
+        # June 2024.
+        # https://github.com/r-lib/cpp11/pull/78
+        # https://www.rstudio.com/about/platform-support/
+        # Ubuntu 20.04 technically dropped support for gcc 4.8, so we have to
+        # add old archives back in manually to install it
+        # https://github.com/r-lib/cpp11/pull/279
         if: matrix.config.custom == 'gcc 4.8'
         run: |
+          echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial main" | sudo tee -a /etc/apt/sources.list
+          echo "deb http://dk.archive.ubuntu.com/ubuntu/ xenial universe" | sudo tee -a /etc/apt/sources.list
+          sudo apt update
           sudo apt-get install -y g++-4.8
           mkdir ~/.R/
-          echo $'CXX1X=g++-4.8\nCXX11=g++-4.8' >> ~/.R/Makevars
+          echo $'CXX1X=g++-4.8\nCXX11=g++-4.8\nCC=gcc-4.8' >> ~/.R/Makevars
```

Comment:
The addition of `CC` is fairly tricky. If you don't add this, the FAQ vignette fails to build with errors about `-fdiagnostics-color=always` being a flag that isn't supported. Note that the FAQ vignette has `{cpp11}` knitr chunks that it compiles, and while it was compiling those chunks it was indeed setting the `-fdiagnostics-color=always` flag.

Some research turned up that this flag was only supported in gcc >=4.9, so gcc is right that it shouldn't be there.

It turns out this flag was being added by pkgbuild during `pkgbuild::build()`.
https://github.com/r-lib/pkgbuild/blob/80762db02d9273997ed6234f654633359269cc9e/R/compiler-flags.R#L51-L58
https://github.com/r-lib/pkgbuild/blob/80762db02d9273997ed6234f654633359269cc9e/R/build.R#L54

It gets added if it looks like the compiler supports colored diagnostics, i.e. the `has_compiler_colored_diagnostics()` check there. This helper works by attempting to compile a simple C program with `-fdiagnostics-color=always` set in the Makevars https://github.com/r-lib/pkgbuild/blob/80762db02d9273997ed6234f654633359269cc9e/R/compiler-flags.R#L76. If the compilation fails, we assume there is no color support. The problem is that it attempts this on a _C_ program which uses a compiler that does support color, but the _C++_ compiler we set through `CXX11` does not.

So my fix for this is to just force the C compiler to also be gcc 4.8 so this helper returns `FALSE`.

Corresponds to this commit https://github.com/r-lib/cpp11/pull/279/commits/f59c83342f4fc9898990e01c332011f16bf3b392

Reply:
yeah, that sounds right, I wonder if pkgbuild should test both the CC and CXX points.

Alternatively perhaps it is better practice to use the same toolchain for C and C++ anyway (even though there is no C code being compiled here)

Reply:
> I wonder if pkgbuild should test both the CC and CXX points.

Gabor and I are talking about exactly that in https://github.com/r-lib/pkgbuild/issues/141